### PR TITLE
Add signal handler to stop gnmi server for when sigterm or sigquit is called

### PR DESCRIPTION
**Pull request by zbud-msft**
#### Why I did it

This PR introduces enhancements to gnmi server such that our server can quickly respond to shutdown signal such as SIGTERM or SIGQUIT. This ensures that our server shutdown is more predictable when SIGTERM or SIGQUIT is called during container restarts.

We use a channel to notify the server of a shutdown signal in a manner that is thread safe and ensures that the server is shut down correctly and necessary clean up steps like log.Flush are executed.

With this PR, we also allow easy transitional changes in the future if we do want to support graceful stops/restarts.

#### How I did it

Added syscall monitoring via a signal channel that is notified when SIGTERM/SIGQUIT is received. This sigchannel, when notified, will update a channel serverControlSignal that is used by our startGNMIServer function in order to handle these syscalls and when it receives a value will stop the gnmi server.

#### How to verify it

UT

#### Which release branch to backport (provide reason below if selected)

- [ ] 201811
- [ ] 201911
- [ ] 202006
- [ ] 202012
- [ ] 202106
- [ ] 202111

### CODE REVIEW

**Comment by zbud-msft** [comment]
Corresponding sonic-mgmt testcase: https://github.com/sonic-net/sonic-mgmt/pull/11702

**Comment by qiluo-msft** [comment]
In your manual test, did you observe the docker stop time reduced? Do you have test result?

**Comment by zbud-msft** [comment]
> In your manual test, did you observe the docker stop time reduced? Do you have test result?

@qiluo-msft With many rounds of manual testing, I am not able to verify that these changes reduce docker stop time. 

Test results:

Without change: 

`time docker stop telemetry`

Average run:

real    0m3.229s
user    0m0.017s
sys     0m0.013s

With change:

`time docker stop telemetry`

Average run:

real    0m3.036s
user    0m0.027s
sys     0m0.025s

Both runs (with and without changes) have a docker stop time of ~3s on average, some runs one or the other might produce a higher number, no consistent proof that these changes reduce docker stop time.

File: telemetry/telemetry.go [hunk 1]
```diff
@@ -149,7 +159,15 @@ func isFlagPassed(fs *flag.FlagSet, name string) bool {
 	return found
 }
 
-func startGNMIServer(telemetryCfg *TelemetryConfig, cfg *gnmi.Config, wg *sync.WaitGroup) {
+func signalHandler(serverControlSignal chan<- int, wg *sync.WaitGroup, sigchannel <-chan os.Signal) {
+	defer wg.Done()
+	select {
+	case <-sigchannel:
+		serverControlSignal <- 0
```
[the remark this answers is not in the document]
Comment (zbud-msft) [reply telemetry/telemetry.go:166]: For future enhancement, we will use this same channel as a way to notify the server if we need to restart if certs have rotated.

File: telemetry/telemetry_test.go [hunk 2]
```diff
@@ -246,7 +270,41 @@ func TestSHA512Checksum(t *testing.T) {
...
+
+	serverControlSignal := make(chan int, 1)
+	testSigChan := make(chan os.Signal, 1)
+	wg := &sync.WaitGroup{}
+
+	wg.Add(1)
+
+	go signalHandler(serverControlSignal, wg, testSigChan)
+
+	testSigChan <- syscall.SIGTERM
```
[the remark this answers is not in the document]
Comment (zbud-msft) [reply telemetry/telemetry_test.go]: Added

**Review by qiluo-msft — APPROVED** [review]

**Pull request merged**